pkg/check/gsoc/gsoc.go

```diff
@@ -0,0 +1,317 @@
+	"github.com/ethersphere/beekeeper/pkg/orchestration"
+	"github.com/gorilla/websocket"
+	"golang.org/x/sync/errgroup"
+)
+
+// Options represents check options
+type Options struct {
+	PostageAmount int64
+	PostageDepth  uint64
+	PostageLabel  string
+}
+
+// NewDefaultOptions returns new default options
+func NewDefaultOptions() Options {
+	return Options{
+		PostageAmount: 1000,
+		PostageDepth:  17,
+		PostageLabel:  "test-label",
+	}
+}
+
+// compile check whether Check implements interface
+var _ beekeeper.Action = (*Check)(nil)
+
+// Check instance.
+type Check struct {
+	logger logging.Logger
+}
+
+// NewCheck returns a new check instance.
+func NewCheck(logger logging.Logger) beekeeper.Action {
+	return &Check{
+		logger: logger,
+	}
+}
+
+func (c *Check) Run(ctx context.Context, cluster orchestration.Cluster, opts interface{}) (err error) {
+	o, ok := opts.(Options)
+	if !ok {
+		return fmt.Errorf("invalid options type")
+	}
+
+	fullNodeNames := cluster.FullNodeNames()
+	clients, err := cluster.NodesClients(ctx)
+	if err != nil {
+		return err
+	}
+
+	if len(fullNodeNames) < 2 {
+		return fmt.Errorf("gsoc test require at least 2 full nodes")
+	}
+
+	uploadClient := clients[fullNodeNames[0]]
+	listenClient := clients[fullNodeNames[1]]
+
+	batches := make([]string, 2)
+	for i := 0; i < 2; i++ {
+		c.logger.Infof("gsoc: creating postage batch. amount=%d, depth=%d, label=%s", o.PostageAmount, o.PostageDepth, o.PostageLabel)
+		batchID, err := uploadClient.CreatePostageBatch(ctx, o.PostageAmount, o.PostageDepth, o.PostageLabel, false)
+		if err != nil {
+			return err
+		}
+		c.logger.Infof("gsoc: postage batch created: %s", batchID)
+		batches[i] = batchID
+	}
+
+	c.logger.Infof("send messages with different postage batches sequentially...")
+	err = run(ctx, uploadClient, listenClient, batches, c.logger, false)
+	if err != nil {
+		return err
+	}
+	c.logger.Infof("done")
+
+	c.logger.Infof("send messages with different postage batches parallel...")
+	err = run(ctx, uploadClient, listenClient, batches, c.logger, true)
+	if err != nil {
+		return err
+	}
+	c.logger.Infof("done")
+
+	return nil
+}
+
+func run(ctx context.Context, uploadClient *bee.Client, listenClient *bee.Client, batches []string, logger logging.Logger, parallel bool) error {
+	const numChunks = 10
+	privKey, err := crypto.GenerateSecp256k1Key()
+	if err != nil {
+		return err
+	}
+
+	addresses, err := listenClient.Addresses(ctx)
+	if err != nil {
+		return err
+	}
+	resourceId, socAddress, err := mineResourceId(ctx, addresses.Overlay, privKey, 1)
+	if err != nil {
+		return err
+	}
+	logger.Infof("gsoc: socAddress=%s, listner node address=%s", socAddress, addresses.Overlay)
+
+	ch, closeWebSocket, err := listenWebsocket(ctx, listenClient.Config().APIURL.Host, socAddress, logger)
+	if err != nil {
+		return fmt.Errorf("listen websocket: %w", err)
+	}
+
+	defer func() {
+		if closeWebSocket != nil {
+			closeWebSocket()
+		}
+	}()
```

Comment:
Maybe it would be cleaner if WebSocket handiling is encapsulated in struct where `Listen` and `Close` methods are exposed. `Close` checks if the WebSocket is already closed to avoid double closures. And then here just call `defer  wsListener.Close()`